### PR TITLE
RAIL-6 Group fares by ticket type code

### PR DESCRIPTION
**Pull request by Dyktus**
All of this to have response like

`"fares": {
            "/journey/44:EDB_KGX_Y70264,KGX_LBG_tube,LBG_TBW_W04675": {
                "/journey/45:TBW_TON_W04751,TON_LBG_W09272,LBG_KGX_tube,KGX_EDB_Y70427": {
                    "/ticket-type/SOR": [
                        "/fare-option/865830368ed6640245b660403f635d3fdd2f7d39"
                    ]
                }
            }
        }`

### CODE REVIEW

File: src/journey/JourneyPlanner.ts [hunk 1]
```diff
@@ -84,22 +84,44 @@ export class JourneyPlanner {
...
+            journeyFareMap[outJourney.id] = {};
             for (const inJourney of inwardPaginatedJourneys) {
                 const inSingle = factory.getSingle(inJourney, fares.inwardSingles);
                 const cheapestReturn = factory.getReturn(outJourney, inJourney, fares.returns);
 
-                const returnPrice = cheapestReturn.map(f => f.totalPrice).getOrElse(Number.MAX_SAFE_INTEGER);
+                let returnPrice = Number.MAX_SAFE_INTEGER;
+                Object.values(cheapestReturn).map(journeys => {
+                    journeys.map((journey) => {
+                        Object.values(journey).map((fareOption: Option<FareOption>) => {
```
Comment (linusnorton) [inline src/journey/JourneyPlanner.ts]: Why are you using map and not returning anything? I think what you want is three for loops

Comment (linusnorton) [reply src/journey/JourneyPlanner.ts]: Also, what is `Object.values(journey)` doing?

File: src/journey/JourneyPlanner.ts [hunk 2]
```diff
@@ -84,22 +84,44 @@ export class JourneyPlanner {
...
             const outSingle = factory.getSingle(outJourney, fares.outwardSingles);
-            journeyFareMap[outJourney.id] = {};
 
+            journeyFareMap[outJourney.id] = {};
             for (const inJourney of inwardPaginatedJourneys) {
                 const inSingle = factory.getSingle(inJourney, fares.inwardSingles);
                 const cheapestReturn = factory.getReturn(outJourney, inJourney, fares.returns);
 
-                const returnPrice = cheapestReturn.map(f => f.totalPrice).getOrElse(Number.MAX_SAFE_INTEGER);
+                let returnPrice = Number.MAX_SAFE_INTEGER;
```
Comment (orzeuek) [inline src/journey/JourneyPlanner.ts]: can you extract `getLowestPrice` method ?

File: src/journey/JourneyPlanner.ts [hunk 3]
```diff
@@ -84,22 +84,44 @@ export class JourneyPlanner {
...
+                });
+                let outSinglePrice = Number.MAX_SAFE_INTEGER;
+                Object.values(outSingle).map(journeys => {
+                    journeys.map((journey) => {
+                        Object.values(journey).map((fareOption: Option<FareOption>) => {
+                            outSinglePrice = fareOption.get.totalPrice < outSinglePrice ? fareOption.get.totalPrice : outSinglePrice;
+                        });
+                    })
+                });
+                let inSinglePrice = Number.MAX_SAFE_INTEGER;
```
Comment (orzeuek) [inline src/journey/JourneyPlanner.ts]: I don't really understand what is he difference between those 3 blocks (return price, out/in single price). Is it possible to extract single method to calculate them? 
Why default value is Number.MAX_SAFE_INTEGER ?
Can you add some comment to explain business requirements behind that logic ?

Comment (orzeuek) [reply src/journey/JourneyPlanner.ts]: Also, I think it's possible to get rid of `let` by replacing map() by reduce()

File: src/service/PersistanceService.ts [hunk 4]
```diff
@@ -92,22 +92,32 @@ export class FareOptionCache {
...
-            if (element instanceof Option) {
-                const option = element as Option<FareOption>;
-                if (option.isEmpty) {
-                    continue;
-                }
-                result.push(option.get);
-            } else if (element instanceof Object || element instanceof Array) {
-                result = [...result, ...this.extractFareOptions(element)]
-            }
+            Object.keys(data[key]).map(journeyKey => {
```
Comment (linusnorton) [inline src/service/PersistanceService.ts]: Why is everything a .map?! If you want to iterate values of an object use `for (const value of Object.values(someObject))` if you want to iterate the keys of an object use `for (const key in someObject)`. If you are iterating an array use `for (const value of someArray)`

File: src/journey/JourneyPlanner.ts [hunk 5]
```diff
@@ -117,11 +139,31 @@ export class JourneyPlanner {
     private getJourneyPlanResponse(outJourneys: Journey[],
                                    inJourneys: Journey[],
                                    journeyFareMap: JourneyFareMap | ReturnJourneyFareMap,
-                                   cache: FareCache,
+                                   cache: {},
                                    paginationResults: PaginationResults): JourneyPlanResponse {
-
-        const outwardJourneys = this.postFilters.reduce((journeys, filter) => filter(cache, journeys), outJourneys);
-        const inwardJourneys = this.postFilters.reduce((journeys, filter) => filter(cache, journeys), inJourneys);
+        let outwardJourneys: Journey[] = [];
```
Comment (orzeuek) [inline src/journey/JourneyPlanner.ts]: Alert is beeping in my mind when I see `let` - in 90% cases you can replace `let` by `const` using some functional programming tricks :)

Comment (orzeuek) [reply src/journey/JourneyPlanner.ts]: 3x `for`
4x `if` 
2x `let` 
please extract function - think about some descriptive name.....

File: src/journey/filter/QuickestOrCheapest.ts [hunk 6]
```diff
@@ -5,7 +5,12 @@ import { FareCache } from '../../validation/JourneyFareOptionFactory';
  * Filter out journeys that are either fastest or cheapest
  */
 export function quickestOrCheapest(fares: FareCache, journeys: Journey[]): Journey[] {
-    const journeysWithFares = journeys.filter(j => fares[j.id].isDefined);
+    const journeysWithFares = journeys.filter(j => {
+        if (typeof fares[j.id] === 'undefined') {
```
Comment (orzeuek) [inline src/journey/filter/QuickestOrCheapest.ts]: how about `if (j.id in fares)` ?

File: src/service/PersistanceService.ts (hunk 4 above)
Comment (orzeuek) [inline src/service/PersistanceService.ts]: I don't really understand this change. That function was recursively walk through whole object (skipping prototype properties) and was looking for object of type `Option` (only FareOptions had this type) - now it doesn't look recursive .....

File: src/service/api/JourneyPlanResponse.ts [hunk 7]
```diff
@@ -63,23 +65,37 @@ export class JourneyPlanResponse {
...
 
+        }
         return result;
     };
 
-    private getFareOptions = (fareOption: Option<FareOption>[]) => {
-        return fareOption.map(fareOptionOption => fareOptionOption.map(this.fareOptionToJSON).orNull);
+    private getFareOptions = (faresGroupedByTicketCode: Option<TicketFares>[]) => {
+        const result: object = {};
+        Object.keys(faresGroupedByTicketCode).map(index => {
```
Comment (orzeuek) [inline src/service/api/JourneyPlanResponse.ts]: sorry, but you are using `.map()` as a replacement of loop. Map mean `transform each element of structure into something different`, `map structure X to structure Y`, in your case it's like foreach.... Sorry man but it's not how world works :) 

If you just want to extract some values from structure, consider usage of `filter` or just `for ... in ...` or  `for ... of ...`

Comment (Dyktus) [reply src/service/api/JourneyPlanResponse.ts]: =) Okey, I used ```for``` at the beginning, but I was also convinced that map is faster. I'll change that

File: src/validation/JourneyFareOptionFactory.ts [hunk 8]
```diff
@@ -45,26 +45,62 @@ class JourneyFareOptionFactoryContext {
...
      * Return the best single FareOption for the given journey and cache the result
      */
-    public getSingle(journey: Journey, fares: SortedFareList): Option<FareOption> {
+    public getSingle(journey: Journey, fares: SortedFareList): Option<TicketFares[]> {
         if (!this.cache[journey.id]) {
             const validator = f => this.validator.isValidOut(f, journey);
             const validSingles = new FaresStream(this.passengerSet, fares, validator);
 
-            this.cache[journey.id] = this.optionFactory(validSingles, this.passengerSet, {});
+            const groups: TicketFares[] = [];
```
Comment (orzeuek) [inline src/validation/JourneyFareOptionFactory.ts]: how about extract `getFaresGroupedByTicketType` method ? or something......

File: src/validation/JourneyFareOptionFactory.ts [hunk 9]
```diff
@@ -45,26 +45,62 @@ class JourneyFareOptionFactoryContext {
...
      * combination is unique.
      */
-    public getReturn(outJourney: Journey, inJourney: Journey, fares: SortedFareList): Option<FareOption> {
+    public getReturn(outJourney: Journey, inJourney: Journey, fares: SortedFareList): Option<TicketFares[]> {
+        const result = {};
         const validator = f => this.validator.isValidOut(f, outJourney) && this.validator.isValidReturn(f, inJourney);
         const validReturns = new FaresStream(this.passengerSet, fares, validator);
 
-        return this.optionFactory(validReturns, this.passengerSet, {});
+        const faresOptions = this.optionFactory(validReturns, this.passengerSet, {});
```
Comment (orzeuek) [inline src/validation/JourneyFareOptionFactory.ts]: see `getSingle` comment

File: src/service/PersistanceService.ts [hunk 10]
```diff
@@ -98,26 +98,16 @@ export class FareOptionCache {
...
-                                    result.push(option.get);
-                                } else if (fareOption instanceof Object || fareOption instanceof Array) {
-                                    result = [...result, ...this.extractFareOptions(fareOption)]
-                                }
-                            });
-                        })
-                    });
-                });
-            });
+            const fareOption: any = data[key];
```
Comment (orzeuek) [inline src/service/PersistanceService.ts]: I changed that in my PR#23 anyway :D

**Pull request merged**